File: Palindrome.js
// A "Palindrome" is a word or sentence which is the same whether read forwards or backwards, for instance "Rise to vote, sir" or "Able was I ere I saw Elba".

// Write a recursive method palindromeRecursive which returns true if a given string is a palindrome. Hint: strings with zero or 1 letters are always palindromes. Note: Do not use the reverse method... too easy!
// Write an iterative method palindromeIterative which does the same iteratively.

var palindromeRecursive=(str)=>{
  if (str.length<2) return true;
  console.log('str[0] = ', str[0], ', str[-1] = ', str[str.length-1])
  if (str[0]!==str[str.length-1]) return false;
  if (str[0]===str[str.length-1]) return palindromeRecursive(str.slice(1, -1))
}
// console.log(palindromeRecursive('yty'), 'should be true')
// console.log(palindromeRecursive('tyuu'), 'should be false')
// console.log(palindromeRecursive('ytu'), 'should be false')
// console.log(palindromeRecursive('ytty'), 'should be true')

var palindromeIterative = (str)=>{
    if (str.length<2) return true;
    for (let i=0; i<str.length; i++) {
      if (str[i] !== str[str.length-(i+1)]){
        return false;
      }
    }
    return true;
}

console.log(palindromeIterative('yty'), 'should be true')
console.log(palindromeIterative('tyuu'), 'should be false')
console.log(palindromeIterative('ytu'), 'should be false')
console.log(palindromeIterative('ytty'), 'should be true')

File: palindrome.rb
# returns true if a given string is a palindrome. Hint: strings with zero or 1 letters are always palindromes. Note: Do not use the reverse method... too easy!

def palindrome_recursive(word)
  if word.length == 1 || word.length == 0
    true
  elsif word[0] == word[-1]
    palindrome_recursive(word[1..-2])
  else
    false
  end
end


def palindrome_iterative(word)
  letters = word.chars
  result = true
  end_i = -1
  letters.each do |letter|
    result = false if letter != word[end_i]
    end_i -= 1
  end
  result
end


input = 'tacocat'
p "Recursive: #{input} is a palindrome? => #{palindrome_recursive(input)}"
p "Iterative: #{input} is a palindrome? => #{palindrome_iterative(input)}"
